### PR TITLE
refactor for new transports

### PR DESCRIPTION
**Pull request by Stebalien**
Removes:

* go-libp2p-conn logic (moved to transports)
* go-peerstream logic (moved here)

Big Changes:

* New transport interface.
* Explicit over implicit (the implicit automagic now all lives in go-libp2p):
  * No more default transports, muxers, etc.
  * No more fallback dialer. Transports are now *required*.
* Stream opening:
  * Connection picking logic (instead of just picking the first).
  * Tries harder to open a stream of some connections happen to be closed.
* Stream closing:
  * No longer treats half-closed streams as fully closed. Users *must* read the an EOF or reset the stream for it to be garbage collected.
* No more polling for dead connections.
* Multiplexers are now *non-optional*. Really, they haven't been optional for a while but we still pretended that they were.
* No more Network type alias. It added a bunch of code and didn't really provide anything but an alternative set of methods that do the same thing.
* Notifications:
  * New guarantee: connection open notifications will complete before connection close notifications begin.
  * Given that, notifications are now delivered in parallel. No more notification backlogs blocking connection closing/opening.

DO NOT MERGE (yet)
Part of: libp2p/go-libp2p#297

This will not currently build (gx deps not updated). However, it works and passes the tests locally.

### CODE REVIEW

File: swarm_addr_test.go [hunk 1]
```diff
@@ -1,100 +1,14 @@
...
 	"testing"
 
-	addrutil "github.com/libp2p/go-addr-util"
-	metrics "github.com/libp2p/go-libp2p-metrics"
 	pstore "github.com/libp2p/go-libp2p-peerstore"
 	testutil "github.com/libp2p/go-testutil"
 	ma "github.com/multiformats/go-multiaddr"
 )
 
-func TestFilterAddrs(t *testing.T) {
```
Comment (Stebalien) [inline swarm_addr_test.go]: The swarm itself now filters addresses based on the registered transports.

File: swarm_dial.go [hunk 2]
```diff
@@ -35,18 +38,24 @@ var (
...
+	// given multiaddr.
+	ErrNoTransport = errors.New("no transport for protocol")
 )
 
-// dialAttempts governs how many times a goroutine will try to dial a given peer.
+// DialAttempts governs how many times a goroutine will try to dial a given peer.
 // Note: this is down to one, as we have _too many dials_ atm. To add back in,
 // add loop back in Dial(.)
-const dialAttempts = 1
+const DialAttempts = 1
```
Comment (Stebalien) [inline swarm_dial.go]: Tests needed access to these and, really, there's no reason not to export them. They're currently constants but we can make them variables later if we want them to be configurable (or even make them swarm parameters).

File: swarm_dial.go [hunk 3]
```diff
@@ -74,7 +83,12 @@ const defaultPerPeerRateLimit = 8
 //  }
 //
 
-type dialbackoff struct {
+// DialBackoff is a type for tracking peer dial backoffs.
+//
+// * It's safe to use it's zero value.
+// * It's thread-safe.
+// * It's *not* safe to move this type after using.
+type DialBackoff struct {
```
Comment (Stebalien) [inline swarm_dial.go]: Exported so we stop exporting a private type from a public function.

File: swarm_dial.go [hunk 4]
```diff
@@ -104,75 +118,81 @@ func (db *dialbackoff) Backoff(p peer.ID) (backoff bool) {
...
-			until: time.Now().Add(baseBackoffTime),
+			until: time.Now().Add(BackoffBase),
 		}
 		return
 	}
 
-	expTimeAdd := time.Second * time.Duration(bp.tries*bp.tries)
-	if expTimeAdd > maxBackoffTime {
-		expTimeAdd = maxBackoffTime
+	backoffTime := BackoffBase + BackoffCoef*time.Duration(bp.tries*bp.tries)
```
Comment (Stebalien) [inline swarm_dial.go]: Nit: Adding in the BackoffBase *here* ensures that we don't go over the maximum (we used to go over by BackoffBase).

File: swarm_dial.go [hunk 5]
```diff
@@ -249,8 +296,10 @@ func (s *Swarm) dial(ctx context.Context, p peer.ID) (*Conn, error) {
 	*/
 	paddrs := s.peers.Addrs(p)
 	goodAddrs := addrutil.FilterAddrs(paddrs,
-		addrutil.AddrUsableFunc,
 		subtractFilter,
+		s.canDial,
+		// TODO: Consider allowing this?
```
Comment (Stebalien) [inline swarm_dial.go]: Eventually, I'd like to allow link-local stuff. For now, this preserves existing behaviour.

File: swarm.go [hunk 6]
```diff
@@ -236,94 +246,180 @@ func (s *Swarm) Close() error {
...
+	default:
+		errCh := make(chan error)
+		for _, c := range conns {
+			go func(c inet.Conn) {
+				errCh <- c.Close()
+			}(c)
 		}
 
-		seen[p] = struct{}{}
+		var errs []string
```
Comment (Stebalien) [inline swarm.go]: Is there a better way to do this? This kind of sucks.

Comment (Stebalien) [reply swarm.go]: Try templates + a wait group.

File: swarm_stream.go [hunk 7]
```diff
@@ -1,69 +1,165 @@
...
-// Stream is a wrapper around a ps.Stream that exposes a way to get
-// our Conn and Swarm (instead of just the ps.Conn and ps.Swarm)
-type Stream ps.Stream
+type streamState int
 
-// Stream returns the underlying peerstream.Stream
-func (s *Stream) Stream() *ps.Stream {
-	return (*ps.Stream)(s)
+const (
+	streamOpen = iota
```
Comment (whyrusleeping) [inline swarm_stream.go]: probs want to use `streamState(iota)`

File: swarm_transport.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+	for _, p := range protocols {
+		transport, ok := s.transports.m[p.Code]
+		if !ok {
+			continue
+		}
+		if transport.Proxy() {
+			return transport
+		}
+	}
+	p := protocols[len(protocols)-1]
```
Comment (magik6k) [inline swarm_transport.go]: Is there any possibility that a multiaddress can specify no protocols? If yes, this may panic.

File: swarm_transport.go [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+import (
+	"fmt"
+	"strings"
+
+	transport "github.com/libp2p/go-libp2p-transport"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// TransportForAddr retrieves the appropriate transport for the given multiaddr.
```
Comment (magik6k) [inline swarm_transport.go]: Would be nice to note that this implements the `Network` interface from `go-libp2p-transport` as it has nice docs

File: swarm_transport.go (hunk 8 above)
Comment (magik6k) [reply swarm_transport.go]: (yes, it would be very weird that such multiaddr would pop-up here but still)

Comment (Stebalien) [reply swarm_transport.go]: There's a check for that at the top of the function (although it shouldn't happen).

File: swarm_transport.go (hunk 9 above)
Comment (Stebalien) [reply swarm_transport.go]: Fair enough.

File: swarm_stream.go (hunk 7 above)
Comment (Stebalien) [reply swarm_stream.go]: Done.

File: swarm_transport.go [hunk 10]
```diff
@@ -26,8 +27,31 @@ func (s *Swarm) TransportForAddr(a ma.Multiaddr) transport.Transport {
...
+// the given multiaddr.
+func (s *Swarm) TransportForListening(a ma.Multiaddr) transport.Transport {
+	protocols := a.Protocols()
+	if len(protocols) == 0 {
+		return nil
+	}
+
+	s.transports.RLock()
+	defer s.transports.RUnlock()
+	selected := s.transports.m[protocols[len(protocols)-1].Code]
```
Comment (vyzo) [inline swarm_transport.go]: will this work for TCP? don't we have `P_TCP` followed by `P_IP` there?

Comment (Stebalien) [reply swarm_transport.go]: Other way around. We have IP followed by TCP. The idea is that, except for proxies like p2p-circuit, each multiaddr protocol "wraps" the previous one. So we pick the last one and then let it deal with the previous ones recursively (possibly delegating back to the swarm or some other transport manager). I'd rather build up from left to right but that starts to be come an abstraction nightmare.

> will

FYI, all this code *does* actually work in practice. I've been running go-ipfs with these changes (all the refactor code) for a while with no issue.

Comment (vyzo) [reply swarm_transport.go]: Yeah, it looks good overall; just getting a little itchy with the possibility of introducing some very subtle bug.

Comment (Stebalien) [reply swarm_transport.go]: I agree. And I know these are a *lot* of changes. That's why I'm letting them bake all together for a while while they're reviewed instead of merging them piece-wise.

File: dial_sync_test.go [hunk 11]
```diff
@@ -8,6 +8,8 @@ import (
 	"time"
 
 	peer "github.com/libp2p/go-libp2p-peer"
+
```
Comment (Kubuxu) [inline dial_sync_test.go]: any reason for a new line here?

File: swarm.go [hunk 12]
```diff
@@ -4,187 +4,169 @@ package swarm
...
 	peers pstore.Peerstore
-	connh ConnHandler
 
-	dsync *DialSync
-	backf dialbackoff
-	dialT time.Duration // mainly for tests
+	conns struct {
+		sync.RWMutex
+		m map[peer.ID][]*Conn
+	}
```
Comment (Kubuxu) [inline swarm.go]: I love this pattern.

File: swarm.go [hunk 13]
```diff
@@ -4,187 +4,169 @@ package swarm
...
+			sync.RWMutex
+			m map[transport.Listener]struct{}
+		}{
+			m: make(map[transport.Listener]struct{}),
+		},
+		transports: struct {
+			sync.RWMutex
+			m map[int]transport.Transport
+		}{
+			m: make(map[int]transport.Transport),
```
Comment (Kubuxu) [inline swarm.go]: I wonder if it would be better to instead do something like this:
```
s.conns.m =  make(map[peer.ID][]*Conn)
```
and so on.
This way we don't have to define those types.

File: swarm.go (hunk 6 above)
Comment (Kubuxu) [reply swarm.go]: Yeah, wait group would be much nicer pattern here.

File: swarm_conn.go [hunk 14]
```diff
@@ -1,122 +1,220 @@
...
+func newConn(s *Swarm, c transport.Conn) *Conn {
+	return &Conn{
+		conn:  c,
+		swarm: s,
+		streams: struct {
+			sync.Mutex
+			m map[*Stream]struct{}
+		}{
+			m: make(map[*Stream]struct{}),
+		},
```
Comment (Kubuxu) [inline swarm_conn.go]: same pattern of map initialization here.

File: swarm.go [hunk 15]
```diff
@@ -4,187 +4,169 @@ package swarm
...
 // be opened and closed, while still using the same Chan for all
 // communication. The Chan sends/receives Messages, which note the
 // destination or source Peer.
-//
-// Uses peerstream.Swarm
 type Swarm struct {
-	swarm *ps.Swarm
+	// Close refcount. This allows us to fully wait for the swarm to be torn
+	// down before continuing.
+	refs sync.WaitGroup
```
Comment (Kubuxu) [inline swarm.go]: I am not sure if WG is right thing to use for this. If Wait was already called and Add is called after it will panic.

File: swarm_stream.go [hunk 16]
```diff
@@ -1,69 +1,165 @@
...
+)
+
+// Stream is the stream type used by swarm. In general, you won't use this type
+// directly.
+type Stream struct {
+	stream smux.Stream
+	conn   *Conn
+
+	stateMu sync.Mutex
+	state   streamState
```
Comment (Kubuxu) [inline swarm_stream.go]: do you want to apply the same anonymous type pattern here?

File: swarm.go (hunk 15 above)
Comment (Stebalien) [reply swarm.go]: Good point, probably easier to read. A waitgroup *should* work as we only call this *after* preventing anything else from starting but yeah, a rwlock is definitely simpler.

File: dial_sync_test.go (hunk 11 above)
Comment (Stebalien) [reply dial_sync_test.go]: import grouping (std, remote, local).

File: swarm.go (hunk 15 above)
Comment (Stebalien) [reply swarm.go]: On further inspection, the code is currently *technically* correct because add only panics if the count reaches zero and then goes above zero). Really, we *want* that behaviour. However, the ref tracking code is still really hard to follow so I'll see what I can do.

Comment (Stebalien) [reply swarm.go]: I tried... there are now more comments. I can't see a nicer (while still efficient) way to do this.

File: dial_sync_test.go (hunk 11 above)
Comment (Stebalien) [reply dial_sync_test.go]: Oh, wait, that's wrong. Should have been std, local, remote (fixed).
